nibabel/gifti/gifti.py

```diff
@@ -834,20 +852,45 @@ def _to_xml_element(self):
             GIFTI.append(dar._to_xml_element())
         return GIFTI
 
-    def to_xml(self, enc='utf-8') -> bytes:
+    def to_xml(self, enc='utf-8', *, mode='strict') -> bytes:
         """Return XML corresponding to image content"""
+        if mode == 'strict':
+            if any(arr.datatype not in GIFTI_DTYPES for arr in self.darrays):
+                raise ValueError(
+                    'GiftiImage contains data arrays with invalid data types; '
+                    'use mode="compat" to automatically cast to conforming types'
+                )
+        elif mode == 'compat':
+            darrays = []
+            for arr in self.darrays:
+                if arr.datatype not in GIFTI_DTYPES:
+                    arr = copy(arr)
+                    # TODO: Better typing for recoders
+                    dtype = cast(np.dtype, data_type_codes.dtype[arr.datatype])
+                    if np.issubdtype(dtype, np.floating):
+                        arr.datatype = data_type_codes['float32']
+                    elif np.issubdtype(dtype, np.integer):
+                        arr.datatype = data_type_codes['int32']
+                    else:
+                        raise ValueError(f'Cannot convert {dtype} to float32/int32')
```

Comment:
I get how this part corrects the darray's `datatype` attribute, but I don't understand where the actual data is casted to the new datatype

Reply:
The data is only cast at write time. Serialization to XML will call `DataArray._to_xml_element()`:

https://github.com/nipy/nibabel/blob/3a4cc5e327c785774b58bf2dee0510bbff9f044f/nibabel/gifti/gifti.py#L487-L522

Which calls `_data_tag_element()`:

https://github.com/nipy/nibabel/blob/3a4cc5e327c785774b58bf2dee0510bbff9f044f/nibabel/gifti/gifti.py#L371-L391

L380 is the one that finally does it: `out = np.asanyarray(dataarray, dtype).tobytes(order)`